kotlinx-coroutines-core/common/src/flow/SharedFlow.kt

```diff
@@ -0,0 +1,652 @@
+/*
+ * Copyright 2016-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.flow.internal.*
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+import kotlin.jvm.*
+import kotlin.native.concurrent.*
+
+/**
+ * A _hot_ [Flow] that shares emitted values among all its collectors in a broadcast fashion, so that all collectors
+ * get all emitted values. A shared flow is called _hot_ because its active instance exists independently of the
+ * presence of collectors. This is opposed to a regular [Flow], such as defined by the [`flow { ... }`][flow] function,
+ * which is _cold_ and is started separately for each collector.
+ *
+ * **Shared flow never completes**. A call to [Flow.collect] on a shared flow never completes normally and
+ * so does a coroutine started by [Flow.launchIn] function. An active collector of a shared flow is called a _subscriber_.
+ *
+ * A subscriber of a shared flow can be cancelled, which usually happens when the scope the coroutine is running
+ * in is cancelled. A subscriber to a shared flow in always [cancellable][Flow.cancellable] and checks for
+ * cancellation before each emission. Note that most terminal operators like [Flow.toList] would not complete, too,
+ * when applied to a shared flow, but flow-truncating operators like [Flow.take] and [Flow.takeWhile] can be used on a
```

Comment:
Might it make sense to have operators such as `toList()` throw an `UnsupportedOperationException` when applied on a `SharedFlow`?

Reply:
No. `SharedFlow` is `Flow` and could be freely used in all context where a `Flow` is expected. Throwing would break encapsulation, leaking flow implementation details to users.